buildkite/data_source_signed_pipeline_steps.go

```diff
@@ -0,0 +1,128 @@
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/lestrrat-go/jwx/v2/jwk"
+	"gopkg.in/yaml.v3"
+)
+
+type signedPipelineStepsDataSource struct {
+	UnisignedSteps types.String `tfsdk:"unsigned_steps"`
+	Repository     types.String `tfsdk:"repository"`
+	JWKS           types.String `tfsdk:"jwks"`
+	JWKSKeyID      types.String `tfsdk:"jwks_key_id"`
+	Steps          types.String `tfsdk:"steps"`
+}
+
+func newSignedPipelineStepsDataSource() datasource.DataSource {
+	return &signedPipelineStepsDataSource{}
+}
+
+func (s *signedPipelineStepsDataSource) Metadata(
+	ctx context.Context,
+	req datasource.MetadataRequest,
+	resp *datasource.MetadataResponse,
+) {
+	resp.TypeName = req.ProviderTypeName + "_signed_pipeline_steps"
+}
+
+func (s *signedPipelineStepsDataSource) Schema(
+	ctx context.Context,
+	req datasource.SchemaRequest,
+	resp *datasource.SchemaResponse,
+) {
+	resp.Schema = schema.Schema{
+		Description: "A data source that can be used to sign pipeline steps with a JWKS key",
+		MarkdownDescription: heredoc.Doc(`
+			Use this data source to look up properties on a specific pipeline. This is particularly useful for looking up the webhook URL for each pipeline.
+
+			More info in the Buildkite [documentation](https://buildkite.com/docs/pipelines).
+		`),
+		Attributes: map[string]schema.Attribute{
+			"unsigned_steps": schema.StringAttribute{
+				Description: "The steps to sign in YAML format.",
+				Required:    true,
+			},
+			"repository": schema.StringAttribute{
+				Description: "The repository that will be checked out in a build of the pipeline.",
+				Required:    true,
+			},
+			"jwks": schema.StringAttribute{
+				Description: "The JWKS to use for signing. If the `jwks_key_id` is not specified, and the set contains exactly one key, that key will be used.",
+				Required:    true,
+			},
+			"jwks_key_id": schema.StringAttribute{
+				Description: "The ID of the key in the JWKS to use for signing.",
+				Required:    false,
+				Optional:    true,
+			},
+				Description: "The signed steps",
+			"steps": schema.StringAttribute{
+				Computed:    true,
+			},
+		},
+	}
+}
+
+func (s *signedPipelineStepsDataSource) Read(
+	ctx context.Context,
+	req datasource.ReadRequest,
+	resp *datasource.ReadResponse,
+) {
+	data := &signedPipelineStepsDataSource{}
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	p, err := pipeline.Parse(strings.NewReader(data.UnisignedSteps.ValueString()))
+	if err != nil {
+		resp.Diagnostics.AddError("Unable to parse pipeline steps", err.Error())
+		return
+	}
+
+	jwks, err := jwk.Parse([]byte(data.JWKS.ValueString()))
+	if err != nil {
+		resp.Diagnostics.AddError("Unable to parse JWKS", err.Error())
+		return
+	}
+
+	var key jwk.Key
+	if data.JWKSKeyID.IsNull() {
+		if jwks.Len() != 1 {
+			resp.Diagnostics.AddError("Cannot find key", "JWKS does not contain exactly one key, but no key ID was specified")
```

Comment:
hmm, we can't use it because it expects to read the key from a file. I did decompose it further, but the method I factored out is private.

Reply:
But the logic in https://github.com/buildkite/go-pipeline/blob/4defc9ce67de6ab876b8c9cd6e4cf34d284a69ae/jwkutil/load_key.go#L56-L76 is:
```
If the keyID is empty 
  If the keySet is a singleton, return the only key in the key set.
  Otherwise, return an error.

If the keyID is not empty, lookup the key by ID in the keySet.
  If the key is found, return the key.
  Otherwise, return an error.
```
I think that the same as here, and it should be fine to use in both situations. LMK if you think otherwise @moskyb.